### PR TITLE
Move fields functions out of the way

### PR DESCRIPTION
**Pull request by jbrockmendel**
Notes:

- The only usage in `tslib` of `get_date_name_field` is for `Timestamp.weekday` and can be simplified with a more direct call.
- Several of the functions from `tslibs.fields` imported into `tslib` are unused.  Updating the external imports to use the new locations is for a follow-up.
- There is some flake8 cleanup to be done in the the new `tslibs.fields`, but that is being saved for a follow-up.  For now this is just cut/paste.
-  `get_start_end_field` will be largely unnecessary if/when the suggestion to deprecate Timestamp.freq in #15146 is adopted.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/tslibs/fields.pyx [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+
+cimport cython
+from cython cimport Py_ssize_t
+
+import numpy as np
+cimport numpy as np
+from numpy cimport ndarray, int64_t, int32_t, int8_t
+np.import_array()
+
+cdef int64_t NPY_NAT = np.datetime64('NaT').astype(np.int64)
```
Comment (jreback) [inline pandas/_libs/tslibs/fields.pyx]: use util.get_nat(), don't change the incantation

File: pandas/_libs/tslibs/fields.pyx [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+        return _isleapyear_arr(get_date_field(dtindex, 'Y'))
+
+    raise ValueError("Field %s not supported" % field)
+
+
+cdef inline int days_in_month(pandas_datetimestruct dts) nogil:
+    return days_per_month_table[is_leapyear(dts.year)][dts.month -1]
+
+
+cpdef _isleapyear_arr(ndarray years):
```
Comment (jreback) [inline pandas/_libs/tslibs/fields.pyx]: can de-privatize

File: pandas/_libs/tslibs/fields.pyx [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+    cdef:
+        ndarray[int8_t] out
+
+    # to make NaT result as False
```
Comment (jreback) [inline pandas/_libs/tslibs/fields.pyx]: I think ok to remove this comment (instead add a doc-string)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some comments. ping on green.

**Comment by jreback** [comment]
> get_start_end_field will be largely unnecessary if/when the suggestion to deprecate Timestamp.freq in #15146 is adopted.

I think this would be a difficult change FYI

File: pandas/_libs/tslibs/fields.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/fields.pyx]: Had hoped to avoid dependency; changing.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**